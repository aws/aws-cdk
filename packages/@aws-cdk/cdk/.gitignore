*.js
*.d.ts
tsconfig.json
tslint.json
*.js.map
dist
coverage
.nyc_output
.jsii

.LAST_BUILD
.nycrc
.LAST_PACKAGE
*.snk