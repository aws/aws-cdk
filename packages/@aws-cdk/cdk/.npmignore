
dist
.LAST_PACKAGE
*.tsbuildinfo
.LAST_BUILD
*.snk
*.ts
!*.d.ts
!*.js
coverage
.nyc_output
*.tgz

# Include .jsii
!.jsii
