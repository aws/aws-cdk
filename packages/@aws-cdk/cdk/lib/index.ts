export * from './aspects/aspect';
export * from './aspects/tag-aspect';

export * from './core/construct';
export * from './core/tokens';
export * from './core/tag-manager';
export * from './core/dependency';

export * from './cloudformation/cloudformation-json';
export * from './cloudformation/cfn-tokens';
export * from './cloudformation/condition';
export * from './cloudformation/fn';
export * from './cloudformation/include';
export * from './cloudformation/logical-id';
export * from './cloudformation/mapping';
export * from './cloudformation/output';
export * from './cloudformation/parameter';
export * from './cloudformation/pseudo';
export * from './cloudformation/resource';
export * from './cloudformation/resource-policy';
export * from './cloudformation/rule';
export * from './cloudformation/stack';
export * from './cloudformation/stack-element';
export * from './cloudformation/dynamic-reference';
export * from './cloudformation/tag';
export * from './cloudformation/removal-policy';
export * from './cloudformation/arn';
export * from './cloudformation/secret';

export * from './app';
export * from './context';
export * from './environment';

export * from './runtime';
