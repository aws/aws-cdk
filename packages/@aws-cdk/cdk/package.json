{
  "name": "@aws-cdk/cdk",
  "version": "0.33.0",
  "description": "AWS Cloud Development Kit Core Library",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "jsii": {
    "outdir": "dist",
    "targets": {
      "java": {
        "package": "software.amazon.awscdk",
        "maven": {
          "groupId": "software.amazon.awscdk",
          "artifactId": "cdk"
        }
      },
      "dotnet": {
        "namespace": "Amazon.CDK",
        "packageId": "Amazon.CDK",
        "signAssembly": true,
        "assemblyOriginatorKeyFile": "../../key.snk"
      },
      "python": {
        "distName": "aws-cdk.cdk",
        "module": "aws_cdk.cdk"
      }
    }
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/awslabs/aws-cdk.git",
    "directory": "packages/@aws-cdk/cdk"
  },
  "awslint": {
    "exclude": [
      "construct-ctor:@aws-cdk/cdk.App.<initializer>",
      "construct-ctor:@aws-cdk/cdk.Root.<initializer>",
      "construct-ctor:@aws-cdk/cdk.Stack.<initializer>.params*",
      "props-no-any:@aws-cdk/cdk.CfnOutputProps.value",
      "props-no-any:@aws-cdk/cdk.CfnParameterProps.default",
      "props-no-any:@aws-cdk/cdk.CfnResourceProps.properties",
      "props-no-cfn-types:@aws-cdk/cdk.CfnOutputProps*",
      "props-no-cfn-types:@aws-cdk/cdk.StringListCfnOutputProps*"
    ]
  },
  "scripts": {
    "build": "cdk-build",
    "watch": "cdk-watch",
    "lint": "cdk-lint",
    "test": "cdk-test",
    "pkglint": "pkglint -f",
    "package": "cdk-package",
    "awslint": "cdk-awslint",
    "build+test+package": "npm run build+test && npm run package",
    "build+test": "npm run build && npm test"
  },
  "nyc": {
    "statements": 55,
    "lines": 55,
    "branches": 35
  },
  "keywords": [
    "aws",
    "cdk",
    "jsii",
    "cdk-core"
  ],
  "author": {
    "name": "Amazon Web Services",
    "url": "https://aws.amazon.com",
    "organization": true
  },
  "license": "Apache-2.0",
  "devDependencies": {
    "@types/lodash": "^4.14.130",
    "cdk-build-tools": "^0.33.0",
    "cfn2ts": "^0.33.0",
    "fast-check": "^1.14.0",
    "lodash": "^4.17.11",
    "pkglint": "^0.33.0"
  },
  "dependencies": {
    "@aws-cdk/cx-api": "^0.33.0"
  },
  "homepage": "https://github.com/awslabs/aws-cdk",
  "peerDependencies": {
    "@aws-cdk/cx-api": "^0.33.0"
  },
  "engines": {
    "node": ">= 8.10.0"
  }
}
