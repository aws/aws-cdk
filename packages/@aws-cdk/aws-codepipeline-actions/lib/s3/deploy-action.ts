import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as s3 from '@aws-cdk/aws-s3';
import { Construct, Duration } from '@aws-cdk/core';
import { kebab as toKebabCase } from 'case';
import { Action } from '../action';
import { deployArtifactBounds } from '../common';

// Class copied verbatim from the aws-s3-deployment module.
// Yes, it sucks that we didn't abstract this properly in a common class,
// but having 2 different CacheControl classes that behave differently would be worse I think.
// Something to do when CDK 2.0.0 comes out.
/**
 * Used for HTTP cache-control header, which influences downstream caches.
 * Use the provided static factory methods to construct instances of this class.
 * Used in the {@link S3DeployActionProps.cacheControl} property.
 *
 * @see https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9
 */
export class CacheControl {
  /** The 'must-revalidate' cache control directive. */
  public static mustRevalidate() { return new CacheControl('must-revalidate'); }
  /** The 'no-cache' cache control directive. */
  public static noCache() { return new CacheControl('no-cache'); }
  /** The 'no-transform' cache control directive. */
  public static noTransform() { return new CacheControl('no-transform'); }
  /** The 'public' cache control directive. */
  public static setPublic() { return new CacheControl('public'); }
  /** The 'private' cache control directive. */
  public static setPrivate() { return new CacheControl('private'); }
  /** The 'proxy-revalidate' cache control directive. */
  public static proxyRevalidate() { return new CacheControl('proxy-revalidate'); }
  /** The 'max-age' cache control directive. */
  public static maxAge(t: Duration) { return new CacheControl(`max-age: ${t.toSeconds()}`); }
  /** The 's-max-age' cache control directive. */
  public static sMaxAge(t: Duration) { return new CacheControl(`s-max-age: ${t.toSeconds()}`); }
  /**
   * Allows you to create an arbitrary cache control directive,
   * in case our support is missing a method for a particular directive.
   */
  public static fromString(s: string) {  return new CacheControl(s); }

  /** @param value the actual text value of the created directive */
  private constructor(public value: string) {}
}

/**
 * Construction properties of the {@link S3DeployAction S3 deploy Action}.
 */
export interface S3DeployActionProps extends codepipeline.CommonAwsActionProps {
  /**
   * Should the deploy action extract the artifact before deploying to Amazon S3.
   *
   * @default true
   */
  readonly extract?: boolean;

  /**
   * The key of the target object. This is required if extract is false.
   */
  readonly objectKey?: string;

  /**
   * The input Artifact to deploy to Amazon S3.
   */
  readonly input: codepipeline.Artifact;

  /**
   * The Amazon S3 bucket that is the deploy target.
   */
  readonly bucket: s3.IBucket;

  /**
   * The specified canned ACL to objects deployed to Amazon S3.
   * This overwrites any existing ACL that was applied to the object.
   *
   * @default - the original object ACL
   */
  readonly accessControl?: s3.BucketAccessControl;

  /**
   * The caching behavior for requests/responses for objects in the bucket.
   * The final cache control property will be the result of joining all of the provided array elements with a comma
   * (plus a space after the comma).
   *
   * @default - none, decided by the HTTP client
   */
  readonly cacheControl?: CacheControl[];
}

/**
 * Deploys the sourceArtifact to Amazon S3.
 */
export class S3DeployAction extends Action {
  private readonly props: S3DeployActionProps;

  constructor(props: S3DeployActionProps) {
    super({
      ...props,
      resource: props.bucket,
      category: codepipeline.ActionCategory.DEPLOY,
      provider: 'S3',
      artifactBounds: deployArtifactBounds(),
      inputs: [props.input],
    });

    this.props = props;
  }

  protected bound(_scope: Construct, _stage: codepipeline.IStage, options: codepipeline.ActionBindOptions):
  codepipeline.ActionConfig {
    // pipeline needs permissions to write to the S3 bucket
    this.props.bucket.grantWrite(options.role);

    // the Action Role also needs to read from the Pipeline's bucket
    options.bucket.grantRead(options.role);

    const acl = this.props.accessControl;
    return {
      configuration: {
        BucketName: this.props.bucket.bucketName,
        Extract: this.props.extract === false ? 'false' : 'true',
        ObjectKey: this.props.objectKey,
        CannedACL: acl ? toKebabCase(acl.toString()) : undefined,
        CacheControl: this.props.cacheControl && this.props.cacheControl.map(ac => ac.value).join(', '),
      },
    };
  }
}
