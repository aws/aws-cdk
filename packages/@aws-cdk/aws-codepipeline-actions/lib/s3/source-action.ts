import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as targets from '@aws-cdk/aws-events-targets';
import * as s3 from '@aws-cdk/aws-s3';
import { Construct } from '@aws-cdk/core';
import { Action } from '../action';
import { sourceArtifactBounds } from '../common';

/**
 * How should the S3 Action detect changes.
 * This is the type of the {@link S3SourceAction.trigger} property.
 */
export enum S3Trigger {
  /**
   * The Action will never detect changes -
   * the Pipeline it's part of will only begin a run when explicitly started.
   */
  NONE = 'None',

  /**
   * CodePipeline will poll S3 to detect changes.
   * This is the default method of detecting changes.
   */
  POLL = 'Poll',

  /**
   * CodePipeline will use CloudWatch Events to be notified of changes.
   * Note that the Bucket that the Action uses needs to be part of a CloudTrail Trail
   * for the events to be delivered.
   */
  EVENTS = 'Events',
}

/**
 * The CodePipeline variables emitted by the S3 source Action.
 */
export interface S3SourceVariables {
  /** The identifier of the S3 version of the object that triggered the build. */
  readonly versionId: string;

  /** The e-tag of the S3 version of the object that triggered the build. */
  readonly eTag: string;
}

/**
 * Construction properties of the {@link S3SourceAction S3 source Action}.
 */
export interface S3SourceActionProps extends codepipeline.CommonAwsActionProps {
  /**
   *
   */
  readonly output: codepipeline.Artifact;

  /**
   * The key within the S3 bucket that stores the source code.
   *
   * @example 'path/to/file.zip'
   */
  readonly bucketKey: string;

  /**
   * How should CodePipeline detect source changes for this Action.
   * Note that if this is S3Trigger.EVENTS, you need to make sure to include the source Bucket in a CloudTrail Trail,
   * as otherwise the CloudWatch Events will not be emitted.
   *
   * @default S3Trigger.POLL
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/log-s3-data-events.html
   */
  readonly trigger?: S3Trigger;

  /**
   * The Amazon S3 bucket that stores the source code
   */
  readonly bucket: s3.IBucket;
}

/**
 * Source that is provided by a specific Amazon S3 object.
 *
 * Will trigger the pipeline as soon as the S3 object changes, but only if there is
 * a CloudTrail Trail in the account that captures the S3 event.
 */
export class S3SourceAction extends Action {
  private readonly props: S3SourceActionProps;

  constructor(props: S3SourceActionProps) {
    super({
      ...props,
      resource: props.bucket,
      category: codepipeline.ActionCategory.SOURCE,
      provider: 'S3',
      artifactBounds: sourceArtifactBounds(),
      outputs: [props.output],
    });

    if (props.bucketKey.length === 0) {
      throw new Error('Property bucketKey cannot be an empty string');
    }

    this.props = props;
  }

  /** The variables emitted by this action. */
  public get variables(): S3SourceVariables {
    return  {
      versionId: this.variableExpression('VersionId'),
      eTag: this.variableExpression('ETag'),
    };
  }

  protected bound(_scope: Construct, stage: codepipeline.IStage, options: codepipeline.ActionBindOptions):
  codepipeline.ActionConfig {
    if (this.props.trigger === S3Trigger.EVENTS) {
      const id = stage.pipeline.node.uniqueId + 'SourceEventRule' + this.props.bucketKey;
      if (this.props.bucket.node.tryFindChild(id)) {
        // this means a duplicate path for the same bucket - error out
        throw new Error(`S3 source action with path '${this.props.bucketKey}' is already present in the pipeline for this source bucket`);
      }
      this.props.bucket.onCloudTrailWriteObject(id, {
        target: new targets.CodePipeline(stage.pipeline),
        paths: [this.props.bucketKey]
      });
    }

    // we need to read from the source bucket...
    this.props.bucket.grantRead(options.role);

    // ...and write to the Pipeline bucket
    options.bucket.grantWrite(options.role);

    return {
      configuration: {
        S3Bucket: this.props.bucket.bucketName,
        S3ObjectKey: this.props.bucketKey,
        PollForSourceChanges: this.props.trigger && this.props.trigger === S3Trigger.POLL,
      },
    };
  }
}
