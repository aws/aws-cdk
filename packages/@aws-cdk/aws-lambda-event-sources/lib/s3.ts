import * as lambda from '@aws-cdk/aws-lambda';
import * as s3 from '@aws-cdk/aws-s3';
import * as notifs from '@aws-cdk/aws-s3-notifications';

export interface S3EventSourceProps {
  /**
   * The s3 event types that will trigger the notification.
   */
  readonly events: s3.EventType[];

  /**
   * S3 object key filter rules to determine which objects trigger this event.
   * Each filter must include a `prefix` and/or `suffix` that will be matched
   * against the s3 object key. Refer to the S3 Developer Guide for details
   * about allowed filter rules.
   */
  readonly filters?: s3.NotificationKeyFilter[];
}

/**
 * Use S3 bucket notifications as an event source for AWS Lambda.
 */
export class S3EventSource implements lambda.IEventSource {
  constructor(readonly bucket: s3.Bucket, private readonly props: S3EventSourceProps) {

  }

  public bind(target: lambda.IFunction) {
    const filters = this.props.filters || [];
    for (const event of this.props.events) {
      this.bucket.addEventNotification(event, new notifs.LambdaDestination(target), ...filters);
    }
  }
}
