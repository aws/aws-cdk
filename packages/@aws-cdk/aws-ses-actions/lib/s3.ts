import * as iam from '@aws-cdk/aws-iam';
import * as kms from '@aws-cdk/aws-kms';
import * as s3 from '@aws-cdk/aws-s3';
import * as ses from '@aws-cdk/aws-ses';
import * as sns from '@aws-cdk/aws-sns';
import * as cdk from '@aws-cdk/core';

/**
 * Construction properties for a S3 action.
 */
export interface S3Props {
  /**
   * The S3 bucket that incoming email will be saved to.
   */
  readonly bucket: s3.IBucket;

  /**
   * The master key that SES should use to encrypt your emails before saving
   * them to the S3 bucket.
   *
   * @default no encryption
   */
  readonly kmsKey?: kms.IKey;

  /**
   * The key prefix of the S3 bucket.
   *
   * @default no prefix
   */
  readonly objectKeyPrefix?: string;

  /**
   * The SNS topic to notify when the S3 action is taken.
   *
   * @default no notification
   */
  readonly topic?: sns.ITopic;
}

/**
 * Saves the received message to an Amazon S3 bucket and, optionally, publishes
 * a notification to Amazon SNS.
 */
export class S3 implements ses.IReceiptRuleAction {
  constructor(private readonly props: S3Props) {
  }

  public bind(rule: ses.IReceiptRule): ses.ReceiptRuleActionConfig {
    // Allow SES to write to S3 bucket
    // See https://docs.aws.amazon.com/ses/latest/DeveloperGuide/receiving-email-permissions.html#receiving-email-permissions-s3
    const keyPattern = this.props.objectKeyPrefix || '';
    const s3Statement = new iam.PolicyStatement({
      actions: ['s3:PutObject'],
      principals: [new iam.ServicePrincipal('ses.amazonaws.com')],
      resources: [this.props.bucket.arnForObjects(`${keyPattern}*`)],
      conditions: {
        StringEquals: {
          'aws:Referer': cdk.Aws.ACCOUNT_ID,
        },
      },
    });
    this.props.bucket.addToResourcePolicy(s3Statement);

    const policy = this.props.bucket.node.tryFindChild('Policy') as s3.BucketPolicy;
    if (policy) { // The bucket could be imported
      rule.node.addDependency(policy);
    } else {
      rule.node.addWarning('This rule is using a S3 action with an imported bucket. Ensure permission is given to SES to write to that bucket.');
    }

    // Allow SES to use KMS master key
    // See https://docs.aws.amazon.com/ses/latest/DeveloperGuide/receiving-email-permissions.html#receiving-email-permissions-kms
    if (this.props.kmsKey && !/alias\/aws\/ses$/.test(this.props.kmsKey.keyArn)) {
      const kmsStatement = new iam.PolicyStatement({
        actions: ['km:Encrypt', 'kms:GenerateDataKey'],
        principals: [ new iam.ServicePrincipal('ses.amazonaws.com')],
        resources: ['*'],
        conditions: {
          Null: {
            'kms:EncryptionContext:aws:ses:rule-name': 'false',
            'kms:EncryptionContext:aws:ses:message-id': 'false',
          },
          StringEquals: {
            'kms:EncryptionContext:aws:ses:source-account': cdk.Aws.ACCOUNT_ID,
          },
        },
      });

      this.props.kmsKey.addToResourcePolicy(kmsStatement);
    }

    return {
      s3Action: {
        bucketName: this.props.bucket.bucketName,
        kmsKeyArn: this.props.kmsKey ? this.props.kmsKey.keyArn : undefined,
        objectKeyPrefix: this.props.objectKeyPrefix,
        topicArn: this.props.topic ? this.props.topic.topicArn : undefined,
      },
    };
  }
}
