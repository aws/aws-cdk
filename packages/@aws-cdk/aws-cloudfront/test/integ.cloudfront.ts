
import s3 = require('@aws-cdk/aws-s3');
import cdk = require('@aws-cdk/cdk');
import cloudfront = require('../lib');

const app = new cdk.App();

const stack = new cdk.Stack(app, 'aws-cdk-cloudfront');

const sourceBucket = new s3.Bucket(stack, 'Bucket');

new cloudfront.CloudFrontWebDistribution(stack, 'MyDistribution', {
  originConfigs: [
    {
      s3OriginSource: {
        s3BucketSource: sourceBucket
      },
      behaviors : [ {isDefaultBehavior: true}]
    }
  ]
 });

app.run();
