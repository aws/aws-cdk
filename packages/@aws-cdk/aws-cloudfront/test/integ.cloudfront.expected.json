{
  "Resources": {
    "Bucket83908E77": {
      "Type": "AWS::S3::Bucket"
    },
    "MyDistributionCFDistributionDE147309": {
      "Type": "AWS::CloudFront::Distribution",
      "Properties": {
        "DistributionConfig": {
          "CacheBehaviors": [],
          "DefaultCacheBehavior": {
            "AllowedMethods": [
              "GET",
              "HEAD"
            ],
            "CachedMethods": [
              "GET",
              "HEAD"
            ],
            "ForwardedValues": {
              "Cookies": {
                "Forward": "none"
              },
              "QueryString": false
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https"
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "Bucket83908E77",
                  "RegionalDomainName"
                ]
              },
              "Id": "origin1",
              "S3OriginConfig": {}
            }
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": {
            "CloudFrontDefaultCertificate": true
          }
        }
      }
    }
  }
}
